Revised Belief Propagation

jack-dunham opened the pull request:
This PR express belief propagation in terms of the new interface based on [AlgorithmsInterface.jl](https://juliamanifolds.github.io/AlgorithmsInterface.jl/dev/) and the included AlgorithmsInterfaceExtensions.jl library.

JoeyT1994 committed:
Working BP Commit

JoeyT1994 committed:
BP Code

jack-dunham committed:
Express BP in terms of `SweepIterator` interface

Introduce `BeliefPropagationProblem` wrapper to hold the cache and the
error `diff` field.

Also simplifies some kwargs wrangling.

jack-dunham committed:
Add method for `setmessages!` that allows messages from one cache to be set from another cache

jack-dunham committed:
Network is now passed to `forest_cover_edge_sequence` directly.

jack-dunham committed:
test file formatting

jack-dunham committed:
Add `DataGraphsPartitionedGraphsExt` glue for `TensorNetwork` type

Also includes some fixes to the way `TensorNetwork` types are
constructed based on index structure.

jack-dunham committed:
Make abstract tensor network interface more generic.

jack-dunham committed:
BP Caching overhauls

jack-dunham committed:
Remove dead deps

jack-dunham committed:
Fix merge

jack-dunham committed:
Fix type inference in TensorNetwork construction

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jack-dunham committed:
Remove `ITensorBase` dep

jack-dunham committed:
`forest_cover_edge_sequence` now constructs a temporary `NamedGraph` instead of trying to operate on existing graphs

The reason for this is:
- One only cares about the edges of the input graph
- A simple graph cannot be used as it "forgets" its edge names resulting
in recursion
- As shown with `TensorNetwork`, removing edges may not always be
defined.

mtfishman committed:
Sweeping algorithms based on AlgorithmsInterface.jl (#30)

mtfishman committed:
Upgrade to NamedDimsArrays.jl v0.11 (#38)

jack-dunham committed:
[LazyNamedDimsArrays] Fix `parenttype` method

jack-dunham committed:
BP Cache now uses new `DataGraphs`interface

jack-dunham committed:
Adjust `default_message` to take a `message` type as its first argument

jack-dunham committed:
Remove unnecessary code and fix ambiguities in `AbstractTensorNetwork`

jack-dunham committed:
`TensorNetwork` type now uses new DataGraphs interface

jack-dunham committed:
[LazyNamedDimsArrays] New `symnameddims` method that pulls out indices from an array.

jack-dunham committed:
The function `region_scalar` should now return a scalar, rather than a order-0 array

jack-dunham committed:
Fix double counting in `edge_scalars` function

This was caused by the change to the `cache` being backed by a directed
graph.

jack-dunham committed:
Minor code formatting

jack-dunham committed:
Expressed belief propagation in terms of AlgorithmsInterface

jack-dunham committed:
Fixes to TensorNetwork construction from tensor list

jack-dunham committed:
Minor simplifications to `contract_network` interface.

jack-dunham committed:
Upgrade DataGraphs and NamedGraphs dependencies

jack-dunham committed:
[AlgorithmsInterfaceExtensions] Allowing mapping over a generic iterable when constructing nested algorithms

jack-dunham committed:
Upgrade serial BP to use own `<:Algorithm` structs.

jack-dunham committed:
Simplify BP cache to only store factors

jack-dunham committed:
Merge branch 'main' into bp

jack-dunham committed:
Upgrade to DataGraphs v0.3.1 and NamedGraphs v0.10

jack-dunham committed:
Fix compat

jack-dunham committed:
Fix broken merge

Fix broken merge

jack-dunham committed:
Bug fix; upgrade tests

jack-dunham committed:
Add 2D TN test

jack-dunham committed:
Formatting

jack-dunham committed:
Merge branch 'main' into bp

JoeyT1994 commented on test/test_beliefpropagation.jl:
```diff
@@ -0,0 +1,90 @@
+        @assert U * LinearAlgebra.diagm(S) * V ≈ W
+        id = [1.0 0.0; 0.0 1.0]
+        set!(sqrt_Ws, e, id)
+        set!(sqrt_Ws, reverse(e), U * LinearAlgebra.diagm(S) * V)
```
Currently the SVD is not really playing a role here. 
`set!(sqrt_Ws, reverse(e), W)` would work. Or, probably best, take the square root via the SVD and set `U*sqrt(S)` on `e` and `sqrt(S)*V` on `reverse(e)`.

jack-dunham commented on test/test_beliefpropagation.jl:
```diff
@@ -0,0 +1,90 @@
+    end
+
+    bpc = BeliefPropagationCache(tn)
+    bpc = ITensorNetworksNext.beliefpropagation(bpc; maxiter = 1)
```
This is indeed how it functions currently. In`select_algorithm(::typeof(beliefpropagation)...)` the `maxiter` kwarg as default `maxiter = is_tree(cache) ? 1 : nothing`

jack-dunham committed:
Simplify BP code

jack-dunham committed:
Add spin ice test

jack-dunham committed:
Version Bump

jack-dunham committed:
Use `abs2` in message diff function.

jack-dunham committed:
Add method for setting intitial messages; improve spin ice tests.

jack-dunham committed:
Remove redundant `default_message_diff_function` function.

jack-dunham commented on test/test_beliefpropagation.jl:
```diff
@@ -0,0 +1,84 @@
+        tn = spin_ice_tensornetwork(g)
+
+        bpc = ITensorNetworksNext.BeliefPropagationCache(tn)
+        bpc = ITensorNetworksNext.beliefpropagation(bpc; maxiter = 1)
```
I've improved this test by initializing messages using `rand` instead `ones`, and now it takes maybe ~5 iterations to converge to the analytical result, so I've added a `maxiter` and `tol`. Thanks!

jack-dunham committed:
Upgrade to DataGraphs and NamedGraphs to 0.4 and 0.11

jack-dunham committed:
Merge branch 'bp' of https://github.com/jack-dunham/ITensorNetworksNext.jl into bp

jack-dunham committed:
Merge branch 'main' into bp

jack-dunham committed:
Formatting

mtfishman commented:
Partially as a reminder to myself, here are the plans for this PR based on offline discussions:

- Inline `updated_message` into SimpleMessageUpdate `solve!` (maybe bring it back later as needed).
- Requires https://github.com/ITensor/DataGraphs.jl/pull/94, changes there to `similar_graph` will need to be reflected here.

jack-dunham committed:
Upgrade to simplified `similar_graph`

jack-dunham committed:
Remove edge arg in `similar_graph`.

jack-dunham committed:
Inline message computation into `solve!`; use type instead of alg string.

jack-dunham committed:
Add in `PartitionedGraphs` interface methods for `TensorNetwork` and `BeliefPropagationCache`.

mtfishman commented on src/beliefpropagation/beliefpropagationproblem.jl:
```diff
@@ -0,0 +1,244 @@
+
+@kwdef mutable struct StopWhenConvergedState{Iterate, Delta <: Real} <:
+    AI.StoppingCriterionState
+    delta::Delta = NaN
```
Agreed, NaN is maybe slightly too clever for this. Inf makes more sense.

mtfishman commented on src/abstracttensornetwork.jl:
```diff
@@ -1,27 +1,23 @@
-    return tn
-end
+# Need to be careful about removing edges from tensor networks in case there is a bond
+Graphs.rem_edge!(::AbstractTensorNetwork, edge) = not_implemented()
```
This is an interesting case... The only ways I can think of doing this generically are:

1. Reroute the link indices from that edge along a different path through the network, however that relies on there being another path through the network so would only work on non-tree graphs and more specifically if the edge being removed is a [bridge](https://en.wikipedia.org/wiki/Bridge_(graph_theory)).
2. If the link index is dimension 1 in theory you might be able to project it out, but that gets very subtle if you have symmetries and may not even be theoretically possible while preserving the symmetry (i.e. if the link index carries a non-trivial charge).

But best to leave it unimplemented for now as you did here.

mtfishman replied:
(Both of those should probably be their own functions, i.e. `reroute_edge(tn, edge)` and `project_edge(tn, edge)`, and then maybe in the future `rem_edge!` could call out to those but they might be too specialized for that.)

mtfishman commented on src/tensornetwork.jl:
```diff
@@ -93,3 +116,113 @@ end
+
+function Graphs.connected_components(tn::TensorNetwork)
+    return Graphs.connected_components(underlying_graph(tn))
+end
```
Wouldn't this be a generic NamedGraph function? Why does this need to be implemented here?

mtfishman commented on src/tensornetwork.jl:
```diff
@@ -93,3 +116,113 @@ end
+    return Graphs.connected_components(underlying_graph(tn))
+end
+
+function Graphs.rem_edge!(tn::TensorNetwork, e)
+    if !has_edge(underlying_graph(tn), e)
+        return false
+    end
+    if !isempty(linkinds(tn, e))
+        throw(
+            ArgumentError(
+                "cannot remove edge $e due to tensor indices existing on this edge."
+            )
+        )
+    end
+    rem_edge!(underlying_graph(tn), e)
+    return true
+end
```
I guess this function touches on a deeper question about whether or not we should allow the underlying graph structure to be mismatched from the tensor connectivity... We don't have to tackle that question in this PR (i.e. this PR can just preserve the status quo), but I wanted to bring that up to hear your thoughts on that. I think I've gone back and forth on that question but my feeling now is that we should be very strict about having the underlying graph structure match the tensor connectevity, and I think that would be made easier with the "reverse index lookup" that I brought up in the last meeting.

As a concrete question, is this definition consistent with how things behaved before this PR?

mtfishman replied:
My current feeling would be that we should not define `rem_edge!` for AbstractTensorNetwork/TensorNetwork: it is underdefined, and should be done "manually" in some other way, say with a function that projects onto a certain link index state, reroutes link indices through another path in the network, etc.

mtfishman commented on src/tensornetwork.jl:
```diff
@@ -93,3 +116,113 @@ end
+function NamedGraphs.similar_graph(
+        graph::TensorNetwork,
+        VD::Type,
+        ::Type{<:Nothing},
```
Do we need this one now that we have the version in https://github.com/ITensor/DataGraphs.jl/pull/94 that only takes a single data type input?

mtfishman replied:
I'm not sure how I feel about using Nothing for not specifying a certain data type so I'd like to move away from that if possible...

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+messages(bp_cache::AbstractGraph) = edge_data(bp_cache)
+messages(bp_cache::AbstractGraph, edges) = [message(bp_cache, e) for e in edges]
+
+message(bp_cache::AbstractGraph, edge::AbstractEdge) = messages(bp_cache)[edge]
```
This is a little confusing, it isn't clear if `message` or `messages` is the "core" function. My intuition would be that `message` should be the core function (which should fall back to data graph access of the edge data on the specified edge) and `messages` would be written in terms of `message`.

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+    ms = messages(bp_cache)
+    delete!(ms, edge)
+    return bp_cache
+end
```
Is the idea of this function that it is meant to invalidate the message on the specified edge but not delete the edge? Is there a data graph function for that we could call directly? This definition may be the best one, it just feels a little off having an interface where `messages(bp_cache)` is a mutable view into the messages that we assume implements `delete!(ms, edge)`. If anything I would maybe bias towards not implementing a generic version of `deletemessage!` and expect concrete types to implement it.

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+function setmessage!(bp_cache::AbstractDataGraph, edge, message)
+    setindex!(bp_cache, message, edge)
+    return bp_cache
+end
```
What about:
```julia
function setmessage!(bp_cache::AbstractGraph, edge, message)
   bp_cache[edge] = message
    return bp_cache
end
```
and then it will fail if the input doesn't implement `setindex!`?

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+function setmessage!(bp_cache::QuotientView, edge, message)
+    setmessages!(parent(bp_cache), QuotientEdge(edge), message)
+    return bp_cache
+end
```
It feels a bit off to me that we need to define this, can't we just rely on the definition of `setindex!` for `QuotientView`?

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+        setmessage!(parent(bp_cache), e, message[e])
+    end
+    return bp_cache
+end
```
Same comment as above.

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+factors(bpc::AbstractGraph, vertices::Vector) = [factor(bpc, v) for v in vertices]
+factors(bpc::AbstractGraph{V}, vertex::V) where {V} = factors(bpc, V[vertex])
+
+factor(bpc::AbstractGraph, vertex) = bpc[vertex]
```
Same comments as for `message`/`messages` about be clearer about what the "core" function is.

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+    return bpc
+end
+
+function region_scalar(bp_cache::AbstractGraph, edge::AbstractEdge; alg = "exact")
```
My guess would be that it would be better to do:
```suggestion
function region_scalar(bp_cache::AbstractGraph, edge::AbstractEdge; kwargs...)
```
and then forward the kwargs to `contract_network`.

mtfishman replied:
Also, maybe a better name for this function would be `edge_scalar`, we could think of a generic version that could take edges or vertices later (though maybe that isn't needed and we were just being a bit too clever before).

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+function region_scalar(bp_cache::AbstractGraph, edge::AbstractEdge; alg = "exact")
+    # Make generic to deal with the possibilty of multiple messages.
+    m1s = messages(bp_cache, [edge])
+    m2s = messages(bp_cache, [reverse(edge)])
```
I don't really understand this comment.

mtfishman replied:
More specifically, whether you call `messages(bp_cache, [edge])` or `message(bp_cache, edge)`, isn't there still just one message, but in the first case it is wrapped into a collection of length 1?

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+    state = factors(bp_cache, vertex)
+
+    return contract_network(vcat(messages, state); alg)[]
+end
```
This made me realize: shouldn't this function take a list of vertices in general?

mtfishman replied:
I.e. "region" makes it sound like it would take a list of vertices, and then maybe stick messages around that region and contract it.

mtfishman replied:
Maybe we should rename this to `vertex_scalar` to emphasize it only takes a single vertex/factor, and then keep in mind `region_scalar` as a generalization that takes multiple vertices.

mtfishman commented on test/test_beliefpropagation.jl:
```diff
@@ -0,0 +1,87 @@
+    bpc = ITensorNetworksNext.beliefpropagation(bpc; maxiter = 1)
+    z_bp = scalar(bpc)
+    z_exact = reduce(*, [tn[v] for v in vertices(g)])[]
+    @test z_bp ≈ z_exact atol = 1.0e-14
```
Instead of hardcoding tolerances and element types, can you write the tests in terms of looping over element types?

mtfishman replied:
I think it is nicer to do that from the start rather than later on. We could loop over `elt in (Float32, ComplexF64)`, for example, to test both single precision and complex.

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+setfactor!(bpc::AbstractGraph, vertex, factor) = not_implemented()
+function setfactor!(bpc::AbstractDataGraph, vertex, factor)
+    fs = factors(bpc)
+    setindex!(fs, vertex, factor)
```
```suggestion
    setindex!(fs, factor, vertex)
```
or better yet, `fs[vertex] = factor`.

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+
+function scalar_factors_quotient(bp_cache::AbstractGraph)
+    return vertex_scalars(bp_cache), edge_scalars(bp_cache)
+end
```
I see there isn't really a compelling reason to have this function, we can just call `vertex_scalars(bp_cache)` and `edge_scalars(bp_cache)` directly.

mtfishman commented on src/beliefpropagation/beliefpropagationcache.jl:
```diff
@@ -0,0 +1,174 @@
+end
+function default_message(T::Type, src, dst)
+    array = ones(Tuple(inds(src) ∩ inds(dst)))
+    return convert(T, array)
```
I think this may be slightly better as a code pattern for this:
```suggestion
    return fill!(similar(T, Tuple(inds(src) ∩ inds(dst))), true)
```
There's a subtlety here when dealing with QN conserving and fermionic tensors, but I think this is a reasonable start and we can discuss that later.

An idea I've brought with with Joey is to use a randomized SVD or QR of the factor to get a better guess for the message. Again, not something for this PR but I wanted to bring it up as something to keep in mind.

mtfishman commented on src/beliefpropagation/beliefpropagationproblem.jl:
```diff
@@ -0,0 +1,230 @@
+    return BeliefPropagationSweep(edges) do edge
+        return SimpleMessageUpdate(edge; kwargs...)
+    end
+end
```
Is this needed as a separate function right now?

mtfishman commented on src/LazyNamedDimsArrays/symbolicnameddimsarray.jl:
```diff
@@ -5,6 +5,9 @@ const SymbolicNamedDimsArray{T, N, Parent <: SymbolicArray{T, N}, DimNames} =
 end
+function symnameddims(name, ndarray::AbstractNamedDimsArray)
+    return symnameddims(name, Tuple(inds(ndarray)))
+end
```
Is this needed? It seems like it would be easy enough to write that manually as needed.

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+using NamedGraphs.PartitionedGraphs: QuotientEdge, QuotientView, parent
+
+messages(bp_cache::AbstractGraph) = edge_data(bp_cache)
+messages(bp_cache::AbstractGraph, edges) = [message(bp_cache, e) for e in edges]
```
I think this code pattern is better:
```suggestion
messages(bp_cache::AbstractGraph, edges) = map(e -> message(bp_cache, e), edges)
```
since that allows the output to reflect the input of `edges` better, i.e. if `edges::Indices` it would output a Dictionary, etc.

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,146 @@
+
+factors(bpc::AbstractGraph) = vertex_data(bpc)
+factors(bpc::AbstractGraph, vertices::Vector) = [factor(bpc, v) for v in vertices]
+factors(bpc::AbstractGraph{V}, vertex::V) where {V} = factors(bpc, V[vertex])
```
```suggestion
factors(bpc::AbstractGraph, vertices) = map(v -> factor(bpc, v), vertices)
```
similar to my suggestion about `messages` above. Also, I don't think that `factors(bpc::AbstractGraph{V}, vertex::V) where {V} = factors(bpc, V[vertex])` is a good idea, it is too subtle and potentially ambiguous to rely on matching the vertex type in that way. Fundamentally, it is ambiguous deciding if the input is meant to be single vertex or a set of vertices just based on the type, which is why we have `factor` vs. `factors`.

mtfishman commented on src/AlgorithmsInterfaceExtensions/AlgorithmsInterfaceExtensions.jl:
```diff
@@ -151,8 +151,8 @@ end
 
-function nested_algorithm(f::Function, nalgorithms::Int; kwargs...)
-    return DefaultNestedAlgorithm(f, nalgorithms; kwargs...)
+function nested_algorithm(f::Function, iterable; kwargs...)
+    return DefaultNestedAlgorithm(f, iterable; kwargs...)
 end
```
Is this change deliberate? Wouldn't the `DefaultNestedAlgorithm` constructor need to be changed accordingly?

My guess is we want both, i.e.:
```julia
function nested_algorithm(f::Function, nalgorithms::Int; kwargs...)
    return nested_algorithm(f, Base.OneTo(nalgorithms); kwargs...)
end
function nested_algorithm(f::Function, iterable; kwargs...)
    return DefaultNestedAlgorithm(f, iterable; kwargs...)
end
```
and then `DefaultNestedAlgorithm` would expect an iterable (or maybe AbstractVector?).

jack-dunham committed:
Test BP with differing precisions; remove `atol` test criteria.

jack-dunham committed:
Use `map` instead of comprehension when returning messages.

Co-authored-by: Matt Fishman <mtfishman@users.noreply.github.com>

jack-dunham committed:
Fix `nested_algorithm` methods on iterables.

jack-dunham committed:
Merge branch 'bp' of https://github.com/jack-dunham/ITensorNetworksNext.jl into bp

jack-dunham committed:
Cleanup `AbstractBeliefPropagationCache` interface.

jack-dunham committed:
Remove `Graphs.connected_components` method for `TensorNetwork`

This method was just forwarding the underlying graph.

jack-dunham committed:
Remove unecessary `symnameddims` method.

jack-dunham committed:
Remove confusing code comment.

jack-dunham committed:
Remove `beliefpropagation_sweep` in favour of constructor call.

jack-dunham committed:
Fix message type initialization failing when only factors are provided.

mtfishman commented:
@jack-dunham can you resolve comments that you've addressed so it is easier to track what remains? Also can you rebase on main and format to get tests running cleanly?

jack-dunham committed:
Formatting.

jack-dunham committed:
Remove `edge_data_type` method for `AbstractTensorNetwork`

An `AbstractTensorNetwork` has edge type `Nothing`, which can be
obtained from the `AbstractDataGraph` method.

jack-dunham committed:
Add some tests for `TensorNetwork` type.

jack-dunham committed:
Bug fixes; more tests

mtfishman commented on src/beliefpropagation/abstractbeliefpropagationcache.jl:
```diff
@@ -0,0 +1,138 @@
+
+abstract type AbstractBeliefPropagationCache{V, VD, ED} <: AbstractDataGraph{V, VD, ED} end
+
+factor_type(bpc::AbstractBeliefPropagationCache) = typeof(bpc)
```
```suggestion
factor_type(bpc::AbstractBeliefPropagationCache) = factor_type(typeof(bpc))
```
Can you add a regression test for that (and related functions that don't have tests)?

jack-dunham committed:
Using `Inf` instead of `NaN` for delta initialization in `StopWhenConvergedState`.

jack-dunham committed:
Add some basic tests for `PartitionedGraphs` interactions with `TensorNetwork`.

mtfishman commented on src/beliefpropagation/beliefpropagationcache.jl:
```diff
@@ -0,0 +1,174 @@
+        bpc = new{V, VD, ED, E, typeof(digraph)}(digraph, factors, messages)
+
+        for edge in edges(bpc)
+            get!(() -> default_message(bpc, edge), messages, edge)
```
This seems bad, doesn't this modify the input `messages` in-place?

mtfishman replied:
I guess our plan is to get rid of `default_message` anyway so this constructor will change accordingly, hopefully we can just remove this entirely. Probably then we can then just use the default constructor and not define an inner constructor.

jack-dunham replied:
Yeah on my local repo i've removed this along with `default_message`, and it seems nothing of value was lost.

mtfishman commented on src/beliefpropagation/beliefpropagationcache.jl:
```diff
@@ -0,0 +1,174 @@
+    AbstractBeliefPropagationCache{V, VD, ED}
+    underlying_graph::G # we only use this for the edges.
+    factors::Dictionary{V, VD}
+    messages::Dictionary{E, ED}
```
Maybe for simplicity we could use `Dict` here?

mtfishman replied:
I'm fine either way, it should mostly be an implementation detail. Just curious to hear your thoughts on that.

jack-dunham committed:
Add tests via Claude.

jack-dunham committed:
Refine and redistribute generated tests

jack-dunham committed:
Further BP test improvements

jack-dunham committed:
Fix incomplete `sitenames` and `siteaxes` definitions.

jack-dunham committed:
Remove `default_message` and other fixes.

jack-dunham committed:
Fix test imports

jack-dunham committed:
Formatting.

jack-dunham committed:
Fix and test tensor network graph manipulation functions.

jack-dunham committed:
Refactor `BeliefPropagationCache` -> `MessageCache`, remove abstract type; other small changes.

mtfishman commented on src/beliefpropagation/messagecache.jl:
```diff
@@ -0,0 +1,247 @@
+# Specific for graphs
+function messages(all_messages::AbstractGraph)
+    return map(edge -> message(all_messages, edge), edges(all_messages))
+end
```
Can't this be defined as `messages(all_messages::AbstractGraph) = messages(all_messages, edges(all_messages))`?

mtfishman replied:
(Also `all_messages` is slightly strange to me as an argument name for this.)

mtfishman commented on src/beliefpropagation/messagecache.jl:
```diff
@@ -0,0 +1,247 @@
+        end
+    end
+    return digraph
+end
```
Couldn't we just use a conversion, i.e. `NamedDiGraph(graph::AbstractGraph)`, instead of introducing a new function for this?

mtfishman commented on src/beliefpropagation/messagecache.jl:
```diff
@@ -0,0 +1,247 @@
+function MessageCache{ED}(::UndefInitializer, graph::AbstractGraph) where {ED}
+    messages = Dictionary{edgetype(graph), ED}()
+    return MessageCache(messages, graph)
+end
```
Do we need an undef constructor? Can't we just expect they are always made with messages, and it is up to the user to decide what the messages are (i.e. undef or not)? I'd prefer to keep this PR as minimal as possible.

mtfishman commented on src/beliefpropagation/messagecache.jl:
```diff
@@ -0,0 +1,247 @@
+            underlying_graph::NamedDiGraph{V}
+        ) where {MT, V, E}
+        return new{MT, V, E}(messages, underlying_graph)
+    end
```
Can we avoid this kind of inner constructor and just use the default constructors?

mtfishman commented on src/beliefpropagation/beliefpropagationproblem.jl:
```diff
@@ -0,0 +1,226 @@
+    return cache
+end
+
+function beliefpropagation(network::AbstractGraph, messages::Dictionary; kwargs...)
```
```suggestion
function beliefpropagation(network::AbstractGraph, messages; kwargs...)
```
I think it would be better to accept any iterator of messages, then it is up to the `MessageCache` constructor to convert to a Dictionary (but that is just an internal detail).

mtfishman commented on src/beliefpropagation/beliefpropagationproblem.jl:
```diff
@@ -0,0 +1,226 @@
+    return beliefpropagation(network, cache; kwargs...)
+end
+
+function beliefpropagation(network, cache; kwargs...)
```
```suggestion
function beliefpropagation(network::AbstractGraph, cache::MessageCache; kwargs...)
```
reflecting the comment above.

mtfishman commented on src/beliefpropagation/messagecache.jl:
```diff
@@ -0,0 +1,247 @@
+
+struct MessageCache{MT, V, E} <: AbstractDataGraph{V, Nothing, MT}
+    messages::Dictionary{E, MT}
+    underlying_graph::NamedDiGraph{V}
```
Honestly I was picturing the MessageCache would just be a dictionary, and wouldn't even have a graph stored. Can't the graph be determined from the network that is passed around in the problem? Is a graph really needed here? The edges are defined as the keys of the dictionary anyway.

mtfishman commented on src/beliefpropagation/messagecache.jl:
```diff
@@ -0,0 +1,247 @@
+    NamedDiGraph, Vertices, convert_vertextype, parent_graph_indices, to_graph_index
+
+struct MessageCache{MT, V, E} <: AbstractDataGraph{V, Nothing, MT}
+    messages::Dictionary{E, MT}
```
It seems like we can hardcode the edge type so we can remove that type parameter:
```suggestion
    messages::Dictionary{NamedEdge{V, V}, MT}
```

jack-dunham committed:
Hard code edge type in `MessageCache`.

Co-authored-by: Matthew Fishman <mtfishman@users.noreply.github.com>

jack-dunham committed:
Simplify `factors` and `messages` methods on `AbstractGraph`

jack-dunham committed:
Allow a custom stopping criteria input into `beliefpropagation` using the `stopping_criterion` kwarg.

jack-dunham committed:
Merge branch 'bp' of https://github.com/jack-dunham/ITensorNetworksNext.jl into bp

jack-dunham committed:
Remove `MessageCache` undef initializer.

mtfishman commented on src/beliefpropagation/beliefpropagationproblem.jl:
```diff
@@ -0,0 +1,248 @@
+        ::typeof(beliefpropagation);
+        maxiter,
+        tol,
+        stopping_criterion
```
What I was picturing was just having a single keyword argument `stopping_criterion`. Users can either pass an actual stopping criterion object, or a NamedTuple of keyword arguments, i.e.:
```julia
beliefpropagation(network, cache; stopping_criterion = (; maxiter = 10, tol = 1e-4))
beliefpropagation(network, cache; stopping_criterion = StopAfterIteration(10))
```
Then there would be a single processing function that takes a NamedTuple and constructs the proper stopping criterion object, or if a stopping criteria is passed, it returns that object. See the design of how `alg` and `trunc` process NamedTuple into algorithm and truncation objects in MatrixAlgebraKit.

jack-dunham committed:
Rename argument names to be more consistent.

mtfishman commented on src/beliefpropagation/messagecache.jl:
```diff
@@ -0,0 +1,230 @@
+    elseif cache1.messages != cache2.messages
+        return false
+    end
+    return true
```
```suggestion
    return (cache1.underlying_graph == cache2.underlying_graph) && (cache1.messages == cache2.messages)
```

jack-dunham committed:
Simplify `MessageCache` interface.

jack-dunham committed:
Rename `ennvironment_messages` to `incoming_edge_data`.

jack-dunham committed:
Rename `logscalar` to `bethe_free_energy`; remove `scalar`.

Previous names were too generic.

jack-dunham committed:
Add `messagecache(f, edge)` method.

jack-dunham committed:
Inline belief propagation algorithm construction.

jack-dunham committed:
Upgrade to registered `AlgorithmsInterface` version; remove `solve!` overloads.
